lib/Basic/FileSystem.cpp

```diff
@@ -30,6 +36,137 @@ namespace {
+    // Assume we'll be able to write to both a temporary file and to the final
+    // destination if the final destination doesn't exist yet.
+    return true;
+  }
+
+  // Fail early if we can't write to the final destination.
+  if (!fs::can_write(outputPath))
+    return llvm::make_error_code(llvm::errc::operation_not_permitted);
+
+  // Only use a temporary if the output is a regular file. This handles
+  // things like '-o /dev/null'
+  return fs::is_regular_file(status);
+}
+
+/// Attempts to open a temporary file next to \p outputPath, with the intent
+/// that once the file has been written it will be renamed into place.
+///
+/// Helper for swift::atomicallyWritingToFile.
+///
+/// \param[out] openedStream On success, a stream opened for writing to the
+/// temporary file that was just created.
+/// \param outputPath The path to the final output file, which is used to decide
+/// where to put the temporary.
+/// \param openFlags Controls how the output stream will be opened.
+///
+/// \returns The path to the temporary file that was opened, or \c None if the
+/// file couldn't be created.
+static Optional<std::string>
+tryToOpenTemporaryFile(Optional<llvm::raw_fd_ostream> &openedStream,
+                       const StringRef outputPath,
+                       const llvm::sys::fs::OpenFlags openFlags) {
+  namespace fs = llvm::sys::fs;
+
+  // Create a temporary file path.
+  // Insert a placeholder for a random suffix before the extension (if any).
+  // Then because some tools glob for build artifacts (such as clang's own
+  // GlobalModuleIndex.cpp), also append .tmp.
+  SmallString<128> tempPath;
+  const StringRef outputExtension = llvm::sys::path::extension(outputPath);
+  tempPath = outputPath.drop_back(outputExtension.size());
+  tempPath += "-%%%%%%%%";
+  tempPath += outputExtension;
+  tempPath += ".tmp";
+
+  int fd;
+  const unsigned perms = fs::all_read | fs::all_write;
+  std::error_code EC = fs::createUniqueFile(tempPath, fd, tempPath, perms,
+                                            openFlags);
+
+  if (EC) {
+    // Ignore the specific error; the caller has to fall back to not using a
+    // temporary anyway.
+    return None;
+  }
+
+  openedStream.emplace(fd, /*shouldClose=*/true);
+  // Make sure the temporary file gets removed if we crash.
+  llvm::sys::RemoveFileOnSignal(tempPath);
+  return tempPath.str().str();
+}
+
+std::error_code swift::atomicallyWritingToFile(
+    const StringRef outputPath, const bool binaryMode,
+    const llvm::function_ref<void(llvm::raw_pwrite_stream &)> action) {
+  namespace fs = llvm::sys::fs;
+
+  // FIXME: This is mostly a simplified version of
+  // clang::CompilerInstance::createOutputFile. It would be great to share the
+  // implementation.
+  assert(!outputPath.empty());
+
+  llvm::ErrorOr<bool> canUseTemporary = canUseTemporaryForWrite(outputPath);
+  if (std::error_code error = canUseTemporary.getError())
+    return error;
+
+  Optional<std::string> temporaryPath;
```

Comment:
I like how the refactoring is going! When I see a declaration, followed by a block that computes it, I wonder if it works better as a separate function. But I'm not sure I see that factoring here would be comprehensible, even though the code wants it.

Reply:
Yeah. It's *almost* `llvm::ErrorOr<Optional<std::string>>` but trying to work with that type (and trying to name the function that would be extracted) seems worse than any of the alternatives. `tryToOpenTemporaryFile` represents the best compromise I came up with.